### ISSUE TITLE
Interface for NetworkManager

### ISSUE THREAD

**Issue by delfick** [python-sdbus/python-sdbus#1]
Hello,

I currently have some code I wrote a while ago using dbus-next that tries to use network manager to find and connect to open wireless networks. I recently needed to actually use it and it seems quite a bit flaky, so I'm trying to find ways of improving the code.

I've come across your library which looks interesting. I'm wondering what's involved in making an interface for network manager? When I tried to figure out this code in the first place I really struggled to find good documentation, so perhaps I was looking in the wrong place.

I ended up writing https://gist.github.com/delfick/a6c978cbffb5a8a13e23a229be3cb99e and I'm not overly happy with it :p  

Thanks

**Comment by igo95862** [comment]
> I'm wondering what's involved in making an interface for network manager? 

Defining an interface classes that mirrors the D-Bus interfaces the NetworkManager exposes.

> When I tried to figure out this code in the first place I really struggled to find good documentation, so perhaps I was looking in the wrong place.

Did you struggle with [my documentation](https://python-sdbus.readthedocs.io/en/latest/)? What did you find difficult to understand? I recommend the quick start pages.

I had plans for NetworkManager interfaces in the repository. I might accelerate adding them if you are interested.

**Comment by delfick** [comment]
I haven't looked too deeply at your library yet, but I wrote my current code in September 2019 but never really use it. It's today that I realise what I wrote isn't overly reliable, and now it would be nice if it was and I could use it (I'm always on a Mac but I have a project I have on a rPi and I needed to update it and I realised Raspberry Pi OS has network manager on it)

I just remember at the time finding dbus very confusing to figure out, which I feel is why my current code isn't great.

**Comment by delfick** [comment]
Currently on the rPi I use a libiw C extension handed down to me.

**Comment by igo95862** [comment]
> I realise what I wrote isn't overly reliable

My library supports static type checking with [mypy](http://mypy-lang.org/). This should increase the reliability as most improper type usage can be detected.  (given that interfaces have the types properly specified)

> I just remember at the time finding dbus very confusing to figure out, which I feel is why my current code isn't great.

I recommend D-Feet and browse around dbus connections.

> Raspberry Pi OS

There might be a problem with python version. Debian Buster still ships 3.7 but I only tested on 3.8 and 3.9. The 3.7 might actually work but my unit test suit does not work. (because IsolatedAsyncioTestcase was added in 3.8)

You will also need `libsystemd-dev` package as it contains the header files for sd-bus.

**Comment by delfick** [comment]
The reliability seemed to be more around dbus complaining about objects that didn't exist and then it did this weird thing where it started complaining about limits. I didn't copy the whole error but something along the lines of `Connection ":1.15" is not allowed to add more match rules`.

d-feet sounds useful.

If I have to I can compile my own python again (had to for raspbian). But if you're not using any specific python3.8 features, then I can probably get away with Python3.7

**Comment by delfick** [comment]
As for testing, does your library work with something like https://pypi.org/project/python-dbusmock/ ? My current code has zero tests cause I had no idea how I'd actually go about writing some.

**Comment by igo95862** [comment]
> The reliability seemed to be more around dbus complaining about objects that didn't exist and also it started doing this weird thing after a while where it started complaining about limits. I didn't copy the whole error but something along the lines of Connection ":1.15" is not allowed to add more match rules.

Sounds like your code or library failed to remove match rules after it finished listening for a specific signal.

> As for testing, does your library work with something like https://pypi.org/project/python-dbusmock/ ? My current code has zero tests cause I had no idea how I'd actually go about writing some.

In theory yes. (since its all communication happens over dbus) You probably want to spawn a separated D-Bus socket to run your tests. I actually was thinking about exporting the setup code for unit tests to my library API. It sets up a separated dbus socket to run the unit tests. (but when it won't run on 3.7 as it uses IsolatedAsyncioTestcase)

**Comment by delfick** [comment]
yeah, I imagine that is indeed the case. I decided to find better libraries before exploring that though :)

In my tests I use a combination of [pytest](https://docs.pytest.org/en/stable/), [noseOfYeti](https://noseofyeti.readthedocs.io/en/latest/) and [alt-pytest-asyncio](https://github.com/delfick/alt-pytest-asyncio). It works quite well. Though migrating 50k lines of tests in the internal version of [photons](https://photons.delfick.com) from nose-tests to pytests was mighty tedious! But worth it even just for pytest fixtures.

**Comment by igo95862** [comment]
BTW the interface object of python-sdbus can function as a local object. Meaning the connection status is transparent and the object will behave close to how a remote object would. This means if you replicate a mock-up of NetworkManager functionality in the interfaces (which you probably will need to create the unit tests) you can use them for unit tests.

**Comment by delfick** [comment]
oh yeah. Cool. It definitely does look much more testable than using dbus-next!

**Comment by delfick** [comment]
So what's your process in making sure the interfaces you write are accurate? Is it a case of finding the source for NetworkManager or do we reverse engineer from d-feet?

**Comment by igo95862** [comment]
> do we reverse engineer from d-feet?

Yes. Also there is [NetworkManager API reference](https://people.freedesktop.org/~lkundrak/nm-docs/spec.html) which can be useful for understanding how it functions and for setting default parameters.

**Comment by delfick** [comment]
right :)

**Comment by igo95862** [comment]
I am working on code generator that creates the interface classes from the D-Bus Introspection XML. Should make porting existing interfaces much easier.

**Comment by delfick** [comment]
oh that sounds handy! Let's wait till you have that before looking further into NetworkManager :) (no rush, seems libiw runs fine on raspbian Pi OS so I can continue to live with libiw)

**Comment by igo95862** [comment]
[This is the auto created classes from the interface xml files](https://github.com/igo95862/python-sdbus/blob/master/src/sdbus/async_proxies/network_manager.py)

Its a bit of a mess and I will need to create proper documentation.

**Comment by delfick** [comment]
oohh, shiny. That's a great start :)

Also, 3k lines of interface would be a nightmare by hand!

**Comment by igo95862** [comment]
[I documented everything and added helper classes I think](https://python-sdbus.readthedocs.io/en/latest/proxies/network_manager.html) 😞🤮

This is something how the code should look like:

```python
from asyncio import get_event_loop

from sdbus import sd_bus_open_system, set_default_bus
from sdbus.async_proxies.network_manager import (NetworkDeviceWireless,
                                                 NetworkManager)

system_bus = sd_bus_open_system()

set_default_bus(system_bus)


network_manager = NetworkManager()


async def do_stuff() -> None:
    wifi_device_path = await network_manager.get_device_by_ip_iface('wlan0')

    wifi_device = NetworkDeviceWireless(wifi_device_path)

    # Do stuff here
    ...


loop = get_event_loop()
loop.create_task(do_stuff())
loop.run_forever()

```

I think I will put binds in the separated package in the future. 4000 lines file is a bit too excessive.

Would it be possible for you to try it out? I don't use NetworkManager so I have no ability to test it. The NetworkManager binds are only in the git right now. (not on PyPI)

**Comment by delfick** [comment]
Excellent! Thanks so much :)

I have corrective eye surgery on Thursday so how that goes will determine which of the following few weekends I'll find time to have a look :)

Fingers crossed, at least by mid next week :)

> 4000 lines file is a bit too excessive.

agreed!

**Comment by delfick** [comment]
Ok, my eye surgery went really well!!

So I'm trying to install python-sdbus now and I get this

```
(venv) pi@onboarder-pi:~/dbus $ python -m pip install -e python-sdbus/
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Obtaining file:///home/pi/dbus/python-sdbus
Installing collected packages: sdbus
  Running setup.py develop for sdbus
    ERROR: Command errored out with exit status 1:
     command: /home/pi/dbus/venv/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/home/pi/dbus/python-sdbus/setup.py'"'"'; __file__='"'"'/home/pi/dbus/python-sdbus/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps
         cwd: /home/pi/dbus/python-sdbus/
    Complete output (62 lines):
    running develop
    running egg_info
    writing sdbus.egg-info/PKG-INFO
    writing dependency_links to sdbus.egg-info/dependency_links.txt
    writing top-level names to sdbus.egg-info/top_level.txt
    reading manifest file 'sdbus.egg-info/SOURCES.txt'
    writing manifest file 'sdbus.egg-info/SOURCES.txt'
    running build_ext
    building 'sdbus.sd_bus_internals' extension
    gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/home/pi/dbus/venv/include -I/usr/local/include/python3.9 -c src/sdbus/sd_bus_internals.c -o build/temp.linux-armv7l-3.9/src/sdbus/sd_bus_internals.o
    src/sdbus/sd_bus_internals.c: In function ‘SdBusInterface_create_vtable’:
    src/sdbus/sd_bus_internals.c:450:37: warning: implicit declaration of function ‘SD_BUS_METHOD_WITH_NAMES_OFFSET’; did you mean ‘SD_BUS_METHOD_WITH_OFFSET’? [-Wimplicit-function-declaration]
             sd_bus_vtable temp_vtable = SD_BUS_METHOD_WITH_NAMES_OFFSET(
                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                         SD_BUS_METHOD_WITH_OFFSET
    src/sdbus/sd_bus_internals.c:455:13: error: expected expression before ‘,’ token
                 ,
                 ^
    src/sdbus/sd_bus_internals.c:524:37: warning: implicit declaration of function ‘SD_BUS_SIGNAL_WITH_NAMES’; did you mean ‘SD_BUS_METHOD_WITH_OFFSET’? [-Wimplicit-function-declaration]
             sd_bus_vtable temp_vtable = SD_BUS_SIGNAL_WITH_NAMES(
                                         ^~~~~~~~~~~~~~~~~~~~~~~~
                                         SD_BUS_METHOD_WITH_OFFSET
    src/sdbus/sd_bus_internals.c:524:37: error: invalid initializer
    src/sdbus/sd_bus_internals.c: In function ‘SdBusMessage_seal’:
    src/sdbus/sd_bus_internals.c:92:50: warning: implicit declaration of function ‘strerrorname_np’; did you mean ‘strerror_l’? [-Wimplicit-function-declaration]
                              __LINE__, __FUNCTION__, strerrorname_np(-return_int));                             \
                                                      ^~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c:608:5: note: in expansion of macro ‘CALL_SD_BUS_AND_CHECK’
         CALL_SD_BUS_AND_CHECK(sd_bus_message_seal(self->message_ref, 0, 0));
         ^~~~~~~~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c: In function ‘SdBusMessage_dump’:
    src/sdbus/sd_bus_internals.c:619:27: warning: implicit declaration of function ‘sd_bus_message_dump’; did you mean ‘sd_bus_message_skip’? [-Wimplicit-function-declaration]
         CALL_SD_BUS_AND_CHECK(sd_bus_message_dump(self->message_ref, 0, SD_BUS_MESSAGE_DUMP_WITH_HEADER));
                               ^~~~~~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c:88:26: note: in definition of macro ‘CALL_SD_BUS_AND_CHECK’
             int return_int = sd_bus_function;                                                                   \
                              ^~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c:619:69: error: ‘SD_BUS_MESSAGE_DUMP_WITH_HEADER’ undeclared (first use in this function); did you mean ‘SD_BUS_MESSAGE_METHOD_ERROR’?
         CALL_SD_BUS_AND_CHECK(sd_bus_message_dump(self->message_ref, 0, SD_BUS_MESSAGE_DUMP_WITH_HEADER));
                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c:88:26: note: in definition of macro ‘CALL_SD_BUS_AND_CHECK’
             int return_int = sd_bus_function;                                                                   \
                              ^~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c:619:69: note: each undeclared identifier is reported only once for each function it appears in
         CALL_SD_BUS_AND_CHECK(sd_bus_message_dump(self->message_ref, 0, SD_BUS_MESSAGE_DUMP_WITH_HEADER));
                                                                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c:88:26: note: in definition of macro ‘CALL_SD_BUS_AND_CHECK’
             int return_int = sd_bus_function;                                                                   \
                              ^~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c: In function ‘encode_object_path’:
    src/sdbus/sd_bus_internals.c:2533:10: warning: implicit declaration of function ‘sd_bus_object_path_is_valid’; did you mean ‘sd_bus_creds_get_fsgid’? [-Wimplicit-function-declaration]
         if (!sd_bus_object_path_is_valid(prefix_char_ptr))
              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
              sd_bus_creds_get_fsgid
    src/sdbus/sd_bus_internals.c: In function ‘PyInit_sd_bus_internals’:
    src/sdbus/sd_bus_internals.c:2703:75: error: ‘SD_BUS_VTABLE_SENSITIVE’ undeclared (first use in this function); did you mean ‘SD_BUS_VTABLE_START’?
         CALL_PYTHON_INT_CHECK(PyModule_AddIntConstant(m, "DbusSensitiveFlag", SD_BUS_VTABLE_SENSITIVE));
                                                                               ^~~~~~~~~~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c:140:26: note: in definition of macro ‘CALL_PYTHON_INT_CHECK’
             int return_int = py_function;      \
                              ^~~~~~~~~~~
    error: command '/usr/bin/gcc' failed with exit code 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /home/pi/dbus/venv/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/home/pi/dbus/python-sdbus/setup.py'"'"'; __file__='"'"'/home/pi/dbus/python-sdbus/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' develop --no-deps Check the logs for full command output.

```

On both the python3.7 installed by default and a python3.9 I compiled myself.

**Comment by igo95862** [comment]
> Ok, my eye surgery went really well!!

Awesome!

> So I'm trying to install python-sdbus now and I get this

Seems like I target too recent software.

`strerrorname_np` was added in glibc 2.32 which was released on 2020-08-05. Its not even in ubuntu 20.04.

`SD_BUS_MESSAGE_DUMP_WITH_HEADER` was added on 2020-01-23. I think it is in ubuntu 20.04 but on Debian.

`SD_BUS_SIGNAL_WITH_NAMES` was added 2019-02-26. Not in Debian.


`strerrorname_np`  can be replaced with `strerror` which should be available.

`SD_BUS_MESSAGE_DUMP_WITH_HEADER` and `SD_BUS_SIGNAL_WITH_NAMES` could be fixed by supplying more recent sd-bus header files.


I think I will setup a continuous integration environment to test comparability.

**Comment by delfick** [comment]
hehhee, I see. C extensions are hard :)

**Comment by igo95862** [comment]
I think I will remove the `strerrorname_np` in favor of `strerror`.

I will also look in to building the binary wheels instead of only having source based package on PyPI.

**Comment by igo95862** [comment]
@delfick Sorry for the lack of updates.

I managed to figure out how to build [manylinux](https://github.com/pypa/manylinux) binary wheels including the ARM64 ones.

These wheels are statically linked to `libsystemd` meaning there won't be issues with headers or libraries versions. They also don't need any compiling and should be ready to use once installed.

[I compiled x86_64 and ARM64 for python versions 3.7, 3.8, 3.9.](https://pypi.org/project/sdbus/0.8rc2/#files) They are in a release candidate for 0.8.0 for now. `pip install sdbus==0.8rc2` is the command to install.

Would you mind giving them a try?

**Comment by delfick** [comment]
ooh shiny. I've been busy too so low updates is fine :)

I ran the pip install command and apparently ldconfig isn't a thing on Mac

```
Collecting sdbus==0.8rc2
  Downloading sdbus-0.8rc2.tar.gz (51 kB)
     |████████████████████████████████| 51 kB 425 kB/s
    ERROR: Command errored out with exit status 1:
     command: /Users/stephenmoore/venv/sdbus/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/private/var/folders/0l/6kvsbpf56px14b8btjy86hb00000gn/T/pip-install-7jc8tyau/sdbus_3cfe2c69600641d6b27be9f1e491928f/setup.py'"'"'; __file__='"'"'/private/var/folders/0l/6kvsbpf56px14b8btjy86hb00000gn/T/pip-install-7jc8tyau/sdbus_3cfe2c69600641d6b27be9f1e491928f/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /private/var/folders/0l/6kvsbpf56px14b8btjy86hb00000gn/T/pip-pip-egg-info-mna71c1s
         cwd: /private/var/folders/0l/6kvsbpf56px14b8btjy86hb00000gn/T/pip-install-7jc8tyau/sdbus_3cfe2c69600641d6b27be9f1e491928f/
    Complete output (13 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/private/var/folders/0l/6kvsbpf56px14b8btjy86hb00000gn/T/pip-install-7jc8tyau/sdbus_3cfe2c69600641d6b27be9f1e491928f/setup.py", line 63, in <module>
        LIBSYSTEMD_MAJOR, LIBSYSTEMD_MINOR, LIBSYTEMD_PATCH = get_libsystemd_version()
      File "/private/var/folders/0l/6kvsbpf56px14b8btjy86hb00000gn/T/pip-install-7jc8tyau/sdbus_3cfe2c69600641d6b27be9f1e491928f/setup.py", line 36, in get_libsystemd_version
        process = subprocess_run(
      File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 505, in run
        with Popen(*popenargs, **kwargs) as process:
      File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 951, in __init__
        self._execute_child(args, executable, preexec_fn, close_fds,
      File "/usr/local/Cellar/python@3.9/3.9.5/Frameworks/Python.framework/Versions/3.9/lib/python3.9/subprocess.py", line 1821, in _execute_child
        raise child_exception_type(errno_num, err_msg, err_filename)
    FileNotFoundError: [Errno 2] No such file or directory: 'ldconfig'
    ----------------------------------------
WARNING: Discarding https://files.pythonhosted.org/packages/64/f0/ba774855bdc2002ed72e08586cfa05f0fefb65c928794557ee56688049a2/sdbus-0.8rc2.tar.gz#sha256=7760a8eef7bfc868fcd182d54a18005bc624aad86a0ea41e1ab18fdf006b661a (from https://pypi.org/simple/sdbus/) (requires-python:>=3.7). Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
ERROR: Could not find a version that satisfies the requirement sdbus==0.8rc2
ERROR: No matching distribution found for sdbus==0.8rc2
WARNING: You are using pip version 21.0.1; however, version 21.1.3 is available.
You should consider upgrading via the '/Users/stephenmoore/venv/sdbus/bin/python -m pip install --upgrade pip' command.
```

**Comment by igo95862** [comment]
> sdbus-0.8rc2.tar.gz

Well it downloaded source archive because I only compiled only Linux versions.

> ldconfig isn't a thing on Mac

Not familiar with Mac but does dbus and systemd exist on it? Systemd depends on glibc which [does not work on macos](https://stackoverflow.com/questions/67918855/how-to-install-glibc-on-bigsur).

Didn't you have Raspberry Pi? If its model 3 or 4 it should work with binary wheels. There might be an issue pip being too old.

**Comment by delfick** [comment]
oh yeah! of course! I'm dumb (and very distracted by life things)!

Seems to not like my rPi either :(

<details>
<summary>I'm not great with C</summary>

```
 20:11 ──   ./sdbus/bin/python -m pip install sdbus==0.8rc2
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
Collecting sdbus==0.8rc2
  Using cached sdbus-0.8rc2.tar.gz (51 kB)
Building wheels for collected packages: sdbus
  Building wheel for sdbus (setup.py) ... error
  ERROR: Command errored out with exit status 1:
   command: /home/pi/sdbus/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-bjw9i1u9/sdbus_d98efa50be794ee6a962833ab063cf8e/setup.py'"'"'; __file__='"'"'/tmp/pip-install-bjw9i1u9/sdbus_d98efa50be794ee6a962833ab063cf8e/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-1qhcs32j
       cwd: /tmp/pip-install-bjw9i1u9/sdbus_d98efa50be794ee6a962833ab063cf8e/
  Complete output (70 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-armv7l-3.7
  creating build/lib.linux-armv7l-3.7/sdbus
  copying src/sdbus/dbus_proxy_async.py -> build/lib.linux-armv7l-3.7/sdbus
  copying src/sdbus/__main__.py -> build/lib.linux-armv7l-3.7/sdbus
  copying src/sdbus/interface_generator.py -> build/lib.linux-armv7l-3.7/sdbus
  copying src/sdbus/dbus_proxy_sync.py -> build/lib.linux-armv7l-3.7/sdbus
  copying src/sdbus/__init__.py -> build/lib.linux-armv7l-3.7/sdbus
  copying src/sdbus/sd_bus_internals.py -> build/lib.linux-armv7l-3.7/sdbus
  copying src/sdbus/autodoc.py -> build/lib.linux-armv7l-3.7/sdbus
  copying src/sdbus/dbus_common.py -> build/lib.linux-armv7l-3.7/sdbus
  copying src/sdbus/dbus_exceptions.py -> build/lib.linux-armv7l-3.7/sdbus
  creating build/lib.linux-armv7l-3.7/sdbus_async
  creating build/lib.linux-armv7l-3.7/sdbus_async/dbus_daemon
  copying src/sdbus_async/dbus_daemon/__init__.py -> build/lib.linux-armv7l-3.7/sdbus_async/dbus_daemon
  creating build/lib.linux-armv7l-3.7/sdbus_block
  creating build/lib.linux-armv7l-3.7/sdbus_block/dbus_daemon
  copying src/sdbus_block/dbus_daemon/__init__.py -> build/lib.linux-armv7l-3.7/sdbus_block/dbus_daemon
  copying src/sdbus/py.typed -> build/lib.linux-armv7l-3.7/sdbus
  copying src/sdbus_async/dbus_daemon/py.typed -> build/lib.linux-armv7l-3.7/sdbus_async/dbus_daemon
  copying src/sdbus_block/dbus_daemon/py.typed -> build/lib.linux-armv7l-3.7/sdbus_block/dbus_daemon
  running build_ext
  building 'sdbus.sd_bus_internals' extension
  creating build/temp.linux-armv7l-3.7
  creating build/temp.linux-armv7l-3.7/src
  creating build/temp.linux-armv7l-3.7/src/sdbus
  arm-linux-gnueabihf-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DLIBSYSTEMD_NO_VALIDATION_FUNCS -I/home/pi/sdbus/include -I/usr/include/python3.7m -c src/sdbus/sd_bus_internals.c -o build/temp.linux-armv7l-3.7/src/sdbus/sd_bus_internals.o
  src/sdbus/sd_bus_internals.c: In function ‘SdBusInterface_create_vtable’:
  src/sdbus/sd_bus_internals.c:403:45: warning: implicit declaration of function ‘SD_BUS_METHOD_WITH_NAMES_OFFSET’; did you mean ‘SD_BUS_METHOD_WITH_OFFSET’? [-Wimplicit-function-declaration]
                   sd_bus_vtable temp_vtable = SD_BUS_METHOD_WITH_NAMES_OFFSET(method_name_char_ptr, input_signature_char_ptr, argument_names_char_ptr,
                                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                               SD_BUS_METHOD_WITH_OFFSET
  src/sdbus/sd_bus_internals.c:404:104: error: expected expression before ‘,’ token
                                                                               result_signature_char_ptr, , _SdBusInterface_callback, 0, flags_long);
                                                                                                          ^
  src/sdbus/sd_bus_internals.c:467:45: warning: implicit declaration of function ‘SD_BUS_SIGNAL_WITH_NAMES’; did you mean ‘SD_BUS_METHOD_WITH_OFFSET’? [-Wimplicit-function-declaration]
                   sd_bus_vtable temp_vtable = SD_BUS_SIGNAL_WITH_NAMES(signal_name_char_ptr, signal_signature_char_ptr, signal_args_names_char_ptr, flags_long);
                                               ^~~~~~~~~~~~~~~~~~~~~~~~
                                               SD_BUS_METHOD_WITH_OFFSET
  src/sdbus/sd_bus_internals.c:467:45: error: invalid initializer
  src/sdbus/sd_bus_internals.c: In function ‘SdBusMessage_dump’:
  src/sdbus/sd_bus_internals.c:540:31: warning: implicit declaration of function ‘sd_bus_message_dump’; did you mean ‘sd_bus_message_skip’? [-Wimplicit-function-declaration]
           CALL_SD_BUS_AND_CHECK(sd_bus_message_dump(self->message_ref, 0, SD_BUS_MESSAGE_DUMP_WITH_HEADER));
                                 ^~~~~~~~~~~~~~~~~~~
  src/sdbus/sd_bus_internals.c:80:34: note: in definition of macro ‘CALL_SD_BUS_AND_CHECK’
                   int return_int = sd_bus_function;                                                                                                             \
                                    ^~~~~~~~~~~~~~~
  src/sdbus/sd_bus_internals.c:540:73: error: ‘SD_BUS_MESSAGE_DUMP_WITH_HEADER’ undeclared (first use in this function); did you mean ‘SD_BUS_MESSAGE_METHOD_ERROR’?
           CALL_SD_BUS_AND_CHECK(sd_bus_message_dump(self->message_ref, 0, SD_BUS_MESSAGE_DUMP_WITH_HEADER));
                                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  src/sdbus/sd_bus_internals.c:80:34: note: in definition of macro ‘CALL_SD_BUS_AND_CHECK’
                   int return_int = sd_bus_function;                                                                                                             \
                                    ^~~~~~~~~~~~~~~
  src/sdbus/sd_bus_internals.c:540:73: note: each undeclared identifier is reported only once for each function it appears in
           CALL_SD_BUS_AND_CHECK(sd_bus_message_dump(self->message_ref, 0, SD_BUS_MESSAGE_DUMP_WITH_HEADER));
                                                                           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  src/sdbus/sd_bus_internals.c:80:34: note: in definition of macro ‘CALL_SD_BUS_AND_CHECK’
                   int return_int = sd_bus_function;                                                                                                             \
                                    ^~~~~~~~~~~~~~~
  src/sdbus/sd_bus_internals.c: In function ‘PyInit_sd_bus_internals’:
  src/sdbus/sd_bus_internals.c:2302:79: error: ‘SD_BUS_VTABLE_SENSITIVE’ undeclared (first use in this function); did you mean ‘SD_BUS_VTABLE_START’?
           CALL_PYTHON_INT_CHECK(PyModule_AddIntConstant(m, "DbusSensitiveFlag", SD_BUS_VTABLE_SENSITIVE));
                                                                                 ^~~~~~~~~~~~~~~~~~~~~~~
  src/sdbus/sd_bus_internals.c:128:34: note: in definition of macro ‘CALL_PYTHON_INT_CHECK’
                   int return_int = py_function; \
                                    ^~~~~~~~~~~
  error: command 'arm-linux-gnueabihf-gcc' failed with exit status 1
  ----------------------------------------
  ERROR: Failed building wheel for sdbus
  Running setup.py clean for sdbus
Failed to build sdbus
Installing collected packages: sdbus
    Running setup.py install for sdbus ... error
    ERROR: Command errored out with exit status 1:
     command: /home/pi/sdbus/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-bjw9i1u9/sdbus_d98efa50be794ee6a962833ab063cf8e/setup.py'"'"'; __file__='"'"'/tmp/pip-install-bjw9i1u9/sdbus_d98efa50be794ee6a962833ab063cf8e/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-z9i9ajxb/install-record.txt --single-version-externally-managed --compile --install-headers /home/pi/sdbus/include/site/python3.7/sdbus
         cwd: /tmp/pip-install-bjw9i1u9/sdbus_d98efa50be794ee6a962833ab063cf8e/
    Complete output (70 lines):
    running install
    running build
    running build_py
    creating build
    creating build/lib.linux-armv7l-3.7
    creating build/lib.linux-armv7l-3.7/sdbus
    copying src/sdbus/dbus_proxy_async.py -> build/lib.linux-armv7l-3.7/sdbus
    copying src/sdbus/__main__.py -> build/lib.linux-armv7l-3.7/sdbus
    copying src/sdbus/interface_generator.py -> build/lib.linux-armv7l-3.7/sdbus
    copying src/sdbus/dbus_proxy_sync.py -> build/lib.linux-armv7l-3.7/sdbus
    copying src/sdbus/__init__.py -> build/lib.linux-armv7l-3.7/sdbus
    copying src/sdbus/sd_bus_internals.py -> build/lib.linux-armv7l-3.7/sdbus
    copying src/sdbus/autodoc.py -> build/lib.linux-armv7l-3.7/sdbus
    copying src/sdbus/dbus_common.py -> build/lib.linux-armv7l-3.7/sdbus
    copying src/sdbus/dbus_exceptions.py -> build/lib.linux-armv7l-3.7/sdbus
    creating build/lib.linux-armv7l-3.7/sdbus_async
    creating build/lib.linux-armv7l-3.7/sdbus_async/dbus_daemon
    copying src/sdbus_async/dbus_daemon/__init__.py -> build/lib.linux-armv7l-3.7/sdbus_async/dbus_daemon
    creating build/lib.linux-armv7l-3.7/sdbus_block
    creating build/lib.linux-armv7l-3.7/sdbus_block/dbus_daemon
    copying src/sdbus_block/dbus_daemon/__init__.py -> build/lib.linux-armv7l-3.7/sdbus_block/dbus_daemon
    copying src/sdbus/py.typed -> build/lib.linux-armv7l-3.7/sdbus
    copying src/sdbus_async/dbus_daemon/py.typed -> build/lib.linux-armv7l-3.7/sdbus_async/dbus_daemon
    copying src/sdbus_block/dbus_daemon/py.typed -> build/lib.linux-armv7l-3.7/sdbus_block/dbus_daemon
    running build_ext
    building 'sdbus.sd_bus_internals' extension
    creating build/temp.linux-armv7l-3.7
    creating build/temp.linux-armv7l-3.7/src
    creating build/temp.linux-armv7l-3.7/src/sdbus
    arm-linux-gnueabihf-gcc -pthread -DNDEBUG -g -fwrapv -O2 -Wall -g -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -fPIC -DLIBSYSTEMD_NO_VALIDATION_FUNCS -I/home/pi/sdbus/include -I/usr/include/python3.7m -c src/sdbus/sd_bus_internals.c -o build/temp.linux-armv7l-3.7/src/sdbus/sd_bus_internals.o
    src/sdbus/sd_bus_internals.c: In function ‘SdBusInterface_create_vtable’:
    src/sdbus/sd_bus_internals.c:403:45: warning: implicit declaration of function ‘SD_BUS_METHOD_WITH_NAMES_OFFSET’; did you mean ‘SD_BUS_METHOD_WITH_OFFSET’? [-Wimplicit-function-declaration]
                     sd_bus_vtable temp_vtable = SD_BUS_METHOD_WITH_NAMES_OFFSET(method_name_char_ptr, input_signature_char_ptr, argument_names_char_ptr,
                                                 ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                                 SD_BUS_METHOD_WITH_OFFSET
    src/sdbus/sd_bus_internals.c:404:104: error: expected expression before ‘,’ token
                                                                                 result_signature_char_ptr, , _SdBusInterface_callback, 0, flags_long);
                                                                                                            ^
    src/sdbus/sd_bus_internals.c:467:45: warning: implicit declaration of function ‘SD_BUS_SIGNAL_WITH_NAMES’; did you mean ‘SD_BUS_METHOD_WITH_OFFSET’? [-Wimplicit-function-declaration]
                     sd_bus_vtable temp_vtable = SD_BUS_SIGNAL_WITH_NAMES(signal_name_char_ptr, signal_signature_char_ptr, signal_args_names_char_ptr, flags_long);
                                                 ^~~~~~~~~~~~~~~~~~~~~~~~
                                                 SD_BUS_METHOD_WITH_OFFSET
    src/sdbus/sd_bus_internals.c:467:45: error: invalid initializer
    src/sdbus/sd_bus_internals.c: In function ‘SdBusMessage_dump’:
    src/sdbus/sd_bus_internals.c:540:31: warning: implicit declaration of function ‘sd_bus_message_dump’; did you mean ‘sd_bus_message_skip’? [-Wimplicit-function-declaration]
             CALL_SD_BUS_AND_CHECK(sd_bus_message_dump(self->message_ref, 0, SD_BUS_MESSAGE_DUMP_WITH_HEADER));
                                   ^~~~~~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c:80:34: note: in definition of macro ‘CALL_SD_BUS_AND_CHECK’
                     int return_int = sd_bus_function;                                                                                                             \
                                      ^~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c:540:73: error: ‘SD_BUS_MESSAGE_DUMP_WITH_HEADER’ undeclared (first use in this function); did you mean ‘SD_BUS_MESSAGE_METHOD_ERROR’?
             CALL_SD_BUS_AND_CHECK(sd_bus_message_dump(self->message_ref, 0, SD_BUS_MESSAGE_DUMP_WITH_HEADER));
                                                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c:80:34: note: in definition of macro ‘CALL_SD_BUS_AND_CHECK’
                     int return_int = sd_bus_function;                                                                                                             \
                                      ^~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c:540:73: note: each undeclared identifier is reported only once for each function it appears in
             CALL_SD_BUS_AND_CHECK(sd_bus_message_dump(self->message_ref, 0, SD_BUS_MESSAGE_DUMP_WITH_HEADER));
                                                                             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c:80:34: note: in definition of macro ‘CALL_SD_BUS_AND_CHECK’
                     int return_int = sd_bus_function;                                                                                                             \
                                      ^~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c: In function ‘PyInit_sd_bus_internals’:
    src/sdbus/sd_bus_internals.c:2302:79: error: ‘SD_BUS_VTABLE_SENSITIVE’ undeclared (first use in this function); did you mean ‘SD_BUS_VTABLE_START’?
             CALL_PYTHON_INT_CHECK(PyModule_AddIntConstant(m, "DbusSensitiveFlag", SD_BUS_VTABLE_SENSITIVE));
                                                                                   ^~~~~~~~~~~~~~~~~~~~~~~
    src/sdbus/sd_bus_internals.c:128:34: note: in definition of macro ‘CALL_PYTHON_INT_CHECK’
                     int return_int = py_function; \
                                      ^~~~~~~~~~~
    error: command 'arm-linux-gnueabihf-gcc' failed with exit status 1
    ----------------------------------------
ERROR: Command errored out with exit status 1: /home/pi/sdbus/bin/python -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-bjw9i1u9/sdbus_d98efa50be794ee6a962833ab063cf8e/setup.py'"'"'; __file__='"'"'/tmp/pip-install-bjw9i1u9/sdbus_d98efa50be794ee6a962833ab063cf8e/setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' install --record /tmp/pip-record-z9i9ajxb/install-record.txt --single-version-externally-managed --compile --install-headers /home/pi/sdbus/include/site/python3.7/sdbus Check the logs for full command output.
```
</details>

**Comment by igo95862** [comment]
I think I figured it out. The key is `--only-binary    ':all:'` flag to force pip use binary package.
<details>

```
# podman run -it --arch=arm64 'docker.io/debian:stretch-slim'
$ apt update
$ apt install python3-pip python3-venv
$ python3 --version
Python 3.7.3
$ python3 -m venv venv
$ ./venv/bin/pip install --only-binary ':all:' sdbus
Collecting sdbus
  Downloading sdbus-0.8rc2-cp37-cp37m-manylinux_2_17_aarch64.manylinux2014_aarch64.whl (532 kB)
     |████████████████████████████████| 532 kB 753 kB/s 
Installing collected packages: sdbus
Successfully installed sdbus-0.8rc2
```

</details>

`podman`/`docker` is very useful for debugging.

**Comment by delfick** [comment]
hmm, seems my rPi isn't one of the arch's you've built

```
(pi) ~
 21:02 ──   uname -m
armv7l

(pi) ~
 21:02 ──   cat /proc/device-tree/model
Raspberry Pi 4 Model B Rev 1.2%

(pi) ~
 21:10 ──   ./sdbus/bin/pip install --only-binary ':all:' sdbus
Looking in indexes: https://pypi.org/simple, https://www.piwheels.org/simple
ERROR: Could not find a version that satisfies the requirement sdbus (from versions: none)
ERROR: No matching distribution found for sdbus

```

**Comment by delfick** [comment]
I tried to setup a github actions pipeline the other day for building stuff on an rPi. I got a decent amount there but then got stuck making debs. It's a private repo, I've added you to it if you wanna have a look.

**Comment by igo95862** [comment]
> I tried to setup a github actions pipeline the other day for building stuff on an rPi. I got a decent amount there but then got stuck making debs.

Making debs sucks a lot. (I do have some experience packaging for my work though)

> It's a private repo, I've added you to it if you wanna have a look.

Sure.

I have a local pipeline which should be able to compile `arm7l` target. I will start it right now.

**Comment by igo95862** [comment]
https://github.com/pypa/manylinux/issues/84 aparently `armv7l` image is missing from official manylinux images.

I will have to compile it myself. There are build script included so I hope it won't be that difficult.

**Comment by delfick** [comment]
cool. Thanks!

**Comment by delfick** [comment]
> Making debs sucks a lot. (I do have some experience packaging for my work though)

For the debs I have to make I'm running checkinstall on a buildkite node on my rPi and omg it's so slow. So I wanted to make them a github action. Then I got carried away trying to make nfpm work instead of checkinstall and ran out of time. hahahha

**Comment by igo95862** [comment]
>  I'm running checkinstall on a buildkite node on my rPi and omg it's so slow.

Cross compiling might be a solution. I have some cross compiling setup over here: https://github.com/tada-team/td-libnice-dev

**Comment by delfick** [comment]
oh cool, I'll look at that later. Thanks :)

**Comment by igo95862** [comment]
@delfick There might be an easy way. Booting Raspberry Pi in 64 bit mode: https://www.sindastra.de/p/1345/switch-your-raspberry-pi-to-64-bit-kernel-raspbian/

Raspberry Pi 4 model B has 4 ARM Cortex-A72 cores which support aarch64 instructions. However, this might break some other software.

**Comment by igo95862** [comment]
Actually its better to install full 64 bit image: https://downloads.raspberrypi.org/raspios_arm64/images/raspios_arm64-2021-05-28/

**Comment by delfick** [comment]
yeah. I am confused why it isn't 64bit tbh.

When I find time (probably in 2.5 weeks) I'll reinstall what's on my rPi
and try again :)

**Comment by igo95862** [comment]
> yeah. I am confused why it isn't 64bit tbh.

Raspberry Pi 1, 2, zero are only 32 bit and before Raspberry Pi 4 there was not any device with more than 4 GB of memory.
> When I find time (probably in 2.5 weeks) I'll reinstall what's on my rPi and try again :)

No rush. I guess the Network Manager binds should be ready by that time.

**Comment by delfick** [comment]
> Raspberry Pi 1, 2, zero are only 32 bit and before Raspberry Pi 4 there was not any device with more than 4 GB of memory.

yeah, I was more meaning cause I installed raspberry Pi OS for the first time on my rPi 4, so I'm confused how I didn't end up choosing a 64-bit one!

> No rush. I guess the Network Manager binds should be ready by that time.

wicked

**Comment by igo95862** [comment]
I moved NetworkManager binds to a separated repository: https://github.com/igo95862/python-sdbus-networkmanager

I tested them a little bit and they should be working. [A little example in documentation.](https://python-sdbus-networkmanager.readthedocs.io/en/latest/examples.html)

**Comment by delfick** [comment]
oh cool. Thanks! I'll let you know when I've had a chance to play around with that :)

**Comment by delfick** [comment]
hmm, well now I know why my rPi isn't 64bit. The official imager doesn't offer it as an option. 

https://jamesachambers.com/where-to-get-the-64-bit-raspberry-pi-os-image-for-pi-4-400/

**Comment by delfick** [comment]
So with https://downloads.raspberrypi.org/raspios_arm64/images/raspios_arm64-2021-05-28/2021-05-07-raspios-buster-arm64.zip

I get this:

![Uploading Screen Shot 2021-07-25 at 9.13.23 pm.png…]()

**Comment by igo95862** [comment]
Very strange. Which python version does it have?

**Comment by delfick** [comment]
```/usr/bin/python3.7: ELF 64-bit LSB executable, ARM aarch64, version 1 (SYSV), dynamically linked, interpreter /lib/ld-linux-aarch64.so.1, for GNU/Linux 3.7.0, BuildID[sha1]=bc7c8c338104e94f7336fca96a0b6cc170a22842, stripped```

**Comment by igo95862** [comment]
Maybe try to set sdbus version directly `sdbus==0.8.0`?

Try purging pip cache? `pip cache purge`

**Comment by igo95862** [comment]
Also the `pypi.org/simple` might be the cause.

**Comment by delfick** [comment]
hmm, specifying version and getting rid of the default pip.conf on raspberry pi OS didn't help

```
pi@raspberrypi:~ $ sudo vi /etc/pip.conf
pi@raspberrypi:~ $ cat /etc/pip.conf
# [global]
# extra-index-url=https://www.piwheels.org/simple
pi@raspberrypi:~ $ ./sdbus/bin/python -m pip install --only-binary ':all:' sdbus
Collecting sdbus
  Could not find a version that satisfies the requirement sdbus (from versions: )
No matching distribution found for sdbus
pi@raspberrypi:~ $ ./sdbus/bin/python -m pip install --only-binary ':all:' sdbus==0.8.0
Collecting sdbus==0.8.0
  Could not find a version that satisfies the requirement sdbus==0.8.0 (from versions: )
No matching distribution found for sdbus==0.8.0

```

**Comment by delfick** [comment]
And installing the wheel directly shows something maybe wrong with the wheel?
<img width="1268" alt="Screen Shot 2021-07-25 at 10 24 27 pm" src="https://user-images.githubusercontent.com/109301/126898851-f7d231b3-a7be-4643-8d15-a607c2ae87da.png">

**Comment by delfick** [comment]
I believe the problem is manylinux
<img width="1500" alt="Screen Shot 2021-07-25 at 10 28 16 pm" src="https://user-images.githubusercontent.com/109301/126898933-16f97cee-f3aa-4134-aa75-0a3c95666e16.png">

**Comment by igo95862** [comment]
Hmm. Try updating pip first to latest version...

**Comment by delfick** [comment]
sigh, I should have  thought of that!
<img width="972" alt="Screen Shot 2021-07-25 at 10 42 24 pm" src="https://user-images.githubusercontent.com/109301/126899350-89ccd4a3-5b4b-429a-b9a4-e80694af92f9.png">

Anyway, sleep time now, I play with this tomorrow maybe. Goodnight!

**Comment by igo95862** [comment]
🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉
🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉🎉

**Comment by delfick** [comment]
Your API is so much nicer than dbus-next! my goodness.

Though I've noticed that I get this error

```
2021-07-27 12:49:02,978[ERROR]:asyncio: Exception in callback SdBus.drive()
handle: <Handle SdBus.drive()>
Traceback (most recent call last):
  File "/usr/lib/python3.7/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
SystemError: null argument to internal routine
```

from doing

```
        # device is a NetworkDeviceWireless
        connections = await device.available_connections
        for conn in connections:
            connection = NetworkConnectionSettings(conn, self.system_bus)
            await connection.delete()
```

or say request_scan

```
        await device.request_scan({})
```

Though, something like `await device.get_all_access_points()` works fine.

**Comment by igo95862** [comment]
There is an error with serializng empty dictionary. Nice find.

I will create fix later today.

**Comment by igo95862** [comment]
```
        # device is a NetworkDeviceWireless
        connections = await device.available_connections
        for conn in connections:
            connection = NetworkConnectionSettings(conn, self.system_bus)
            await connection.delete()
```
This bit of code does not cause any issues for me.

**Comment by igo95862** [comment]
I wonder if it might be caused by Python 3.7, old NetworkManager version or something else...

Debugging binary extensions is almost impossible without debug symbols but maybe you will be able to catch the last Python function that gets called.

**Comment by igo95862** [comment]
Also try running unit tests from the repository. Clone it and run `python -m unittest`

**Comment by delfick** [comment]
so, it seems the latest version of network-manager is 1.34 and the one in raspberry pi repositories is 1.14.6 lol

I'll figure out how to get a newer version on my rPi

**Comment by delfick** [comment]
I installed Ubuntu Budgie and I still have that exception happen :(

```
(sdbus) pi@rpi4:~/connector$ uname -a
Linux rpi4 5.11.0-1005-raspi #5-Ubuntu SMP PREEMPT Fri Apr 9 13:10:37 UTC 2021 aarch64 aarch64 aarch64 GNU/Linux
(sdbus) pi@rpi4:~/connector$ python --version
Python 3.9.5
(sdbus) pi@rpi4:~/connector$ nmcli --version
nmcli tool, version 1.30.0
```

**Comment by delfick** [comment]
Still investigating, but it appears that on this line https://github.com/igo95862/python-sdbus/blob/c15b0b3f516b6f189998044596fd2fb6dfd54f54/src/sdbus/sd_bus_internals.c#L1695 that exception_unmapped_message is NULL.

**Comment by delfick** [comment]
I took a break for a bit and when I tried again I did `sd_bus_message_dump(message, stdout, 0);` on line 1676:

and got

```
MESSAGE "s" {
        STRING "org.freedesktop.NetworkManager.wifi.scan request failed: not authorized";
};
```

And proceeded to run with sudo and it no longer throws that error. I'm not convinced request_scan does what I want it to though as the list still shows APs that I've turned off....

**Comment by igo95862** [comment]
Hmm. I see.

Let me try to add test for unmapped error message.

**Comment by igo95862** [comment]
I published version 0.8.1 with the fix.

Thank you for debugging this issue.

**Comment by igo95862** [comment]
Also I guess NetworkManager binds could use some error mappings.

**Comment by delfick** [comment]
Yay, that works, thanks!

> Also I guess NetworkManager binds could use some error mappings.

:)

**Comment by igo95862** [comment]
I have added the errors I found in NetworkManager source code.

They are documented here: https://python-sdbus-networkmanager.readthedocs.io/en/latest/exceptions.html

I also pushed version [1.1rc1](https://pypi.org/project/sdbus-networkmanager/1.1rc1/) with the exceptions.

**Comment by delfick** [comment]
super cool. Looks like that works. Thanks!

**Comment by igo95862** [comment]
I published the 1.1.0 binds version with a common base for all exceptions.

Do you have any feedback?

Would you mind if I close this issue? Maybe we can move to discussions or new issue.

**Comment by delfick** [comment]
yeah, sorry, I was so close to getting done what I wanted with this and then had to stop (started a new job today and decided I should focus more on resting before that started than programming hahhaha). I was gonna finish it and then close this issue. Maybe I'll just close the issue and any further issues when I get around to fixing the last couple bugs in my code I'll make a new one.

Just tried out 1.1.0 and that looks good. Having a base exception is a great idea!

Also, thanks so much for making this happen, such a nicer API to use :)

**Comment by delfick** [comment]
I finally found time to do the final stretch and make sure it works and I think I managed to get it done :)

Final version is here https://github.com/delfick/network_changer/blob/7daee99e6dffaa47a788c07461573d5d57df1eb1/network_changer/platforms/dbus.py

Thanks once again!!
